datalake: metric for invalid records

nvartolomei opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes
* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
datalake: metric for invalid records

nvartolomei committed:
rptest: label values must be strings

andrwng commented on src/v/datalake/translation/translation_probe.h:
```diff
@@ -0,0 +1,51 @@
+    model::ntp _ntp;
+    std::optional<metrics::public_metric_groups> _public_metrics;
+    absl::flat_hash_map<invalid_record_label_set, int64_t>
+      _invalid_record_action_by_label;
```
Do we actually need to separate invalid records based on the action? IMO the motivation of this metric is to signal to users that their topic has invalid records coming in, in which case this could be a simple counter.

Or if it really is important to capture the action in this metric, I'm surprised about the relative complexity of the code, vs two counters and a switch statement.

nvartolomei replied:
The complexity will stay because more important than the action is the cause which I'm planning to add next.

andrwng replied:
Does "cause" necessitate complexity, even? I don't see how "action" and "cause" complexity are coupled. Regardless, why wouldn't the probe just look like:

```
int64_t num_invalid_records;
int64_t num_failed_schema_resolutions;
int64_t num_failed_data_translations;
```

I don't think I understand why the "action" label is useful as a part of the metric. If we want to measure DLQ records, isn't it less hassle to just add a probe to the coordinator?

nvartolomei replied:
Simplified to the level you suggested.

andrwng commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -144,6 +144,15 @@ double datalake_manager::average_translation_backlog() {
 }
 
+ss::lw_shared_ptr<class translation_probe>
+datalake_manager::translation_probe(const model::ntp& ntp) {
```
nit: maybe call this `get_or_create_probe` or something? Was just a bit surprised at the callsite that this was also constructing

andrwng commented on src/v/datalake/translation/translation_probe.h:
```diff
@@ -0,0 +1,51 @@
+namespace datalake {
+
+class translation_probe final {
+    struct invalid_record_label_set {
```
nit: could use docs, but also per my other comment I think I need some convincing that we need this

andrwng commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -178,6 +187,10 @@ ss::future<> datalake_manager::start() {
       = _partition_mgr->local().register_unmanage_notification(
         model::kafka_namespace, [this](model::topic_partition_view tp) {
             model::ntp ntp{model::kafka_namespace, tp.topic, tp.partition};
+            // We remove the probe only when partition is moved out of the
+            // shard to avoid metrics disappearing during much more common
+            // leadership changes.
+            _translation_probe_by_ntp.erase(ntp);
```
Hm I actually thought having the partition probe being the parent was pretty elegant. Curious if you ran up against something that made it impractical? If the concern was about generalizability, I'm sure we could've just put the translation probe into the cluster namespace or something.

Managing the probe in the datalake manager is fine too, but it does feel like there's a bit more overhead in making sure it's correct.

nvartolomei replied:
@andrwng couple of reasons for switching to the current approach:
1) reviewing partition probe (part) was a bit more complex than this and i didn't want to introduce that complexity now; i want to move a little bit faster
2) supporting dynamic label set began to put strain on the api design i began with; mainly because of some peculiarities in the partition probe itself (e.g. clear metrics api, reconfiguration on config change for schema registry) that meant that partition_probe_part needed more API surface or it could manage the metrics group itself but then almost all the benefits of putting it in the partition_probe were lost

I find this design much simpler.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--b7af1e62-0402-4085-aa94-02299b3ced15-->
<!--TEST RESULTS-->
<!--8d8370c9-a454-4d0b-ab07-e03cd49597b7-->
<!--TEST RESULTS-->
<!--1dc75b23-9598-40dd-b084-e687e555fc25-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61742">build#61742</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| partition_balancer_simulator_test_rpunit.partition_balancer_simulator_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61742#0194e205-1c77-4aca-a99e-9610a13d4a2d | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61742#0194e249-db17-4af9-8cdf-67e09fd261d5 | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/61742#0194e264-f397-4e49-b577-eb35e1050b97 | FLAKY | 1/3 |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/61742#0194e249-db17-4d91-9559-5f0e6babef87 | FLAKY | 1/2 |
| rptest.tests.datalake.mount_unmount_test.MountUnmountIcebergTest.test_simple_unmount.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61742#0194e249-db17-4af9-8cdf-67e09fd261d5 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61742#0194e264-f395-4045-8ad6-49f85689be09 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61809">build#61809</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61809#0194f77d-ef11-4a97-938a-8bbb7f358fa1 | FLAKY | 1/3 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61809#0194f77a-2302-47a5-b1bd-c0d96be21af0 | FLAKY | 1/2 |
| rptest.tests.controller_log_limiting_test.ControllerLogLimitMirrorMakerTests.test_mirror_maker_with_limits | ducktape | https://buildkite.com/redpanda/redpanda/builds/61809#0194f77d-ef0e-424e-ab5a-ba113b881731 | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/61809#0194f77d-ef0e-424e-ab5a-ba113b881731 | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=10 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61809#0194f77d-ef11-4a97-938a-8bbb7f358fa1 | FLAKY | 1/4 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=10 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61809#0194f77d-ef0f-4321-a666-8ed19174563f | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61913">build#61913</a></summary>


| test_id                                                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.audit_log_test.AuditLogTestsAppLifecycle.test_app_lifecycle                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61913#01951413-9dd9-42b5-9983-3557d291899a | FLAKY         | 1/2      |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61913#01951418-15f5-4f3a-8bd7-9b58a3a03646 | FLAKY         | 1/2      |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=2.cloud_storage_type=CloudStorageType.ABS                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61913#01951418-15f5-4f3a-8bd7-9b58a3a03646 | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61913#01951418-15f6-42ca-9597-a366fe999abc | FLAKY         | 1/3      |
</details>
<!--/body-->

andrwng commented on src/v/datalake/translation/translation_probe.cc:
```diff
@@ -0,0 +1,82 @@
+
+std::ostream&
+operator<<(std::ostream& os, translation_probe::invalid_record_cause cause) {
+    switch (cause) {
```
nit: might be able to use "using enum" here

andrwng commented on src/v/datalake/translation/translation_probe.h:
```diff
@@ -0,0 +1,71 @@
+            return _num_failed_data_translation;
+        case invalid_record_cause::failed_iceberg_schema_resolution:
+            return _num_failed_iceberg_schema_resolution;
+        default:
+            throw std::invalid_argument(
+              fmt::format("Invalid cause: {}", std::to_underlying(cause)));
```
Isn't it more foolproof to not have a default? I thought that would make the build complain about a missing enum if a user forgets to add something here.

andrwng reviewed (commented):
Thanks for the simplification! Pretty much LGTM, just one note about a switch

nvartolomei committed:
datalake/tests: return schema id on registration

nvartolomei committed:
datalake/tests: fix bad data test

The test before this commit fails finding schema.

Improve the test so that it actually fails with a translation error.

This was caught while testing the metrics addition.

The metrics assertion added later catches the problem.

nvartolomei commented:
* Addressed all comments by @andrwng and updated unit tests to check metric values too.
  * https://github.com/redpanda-data/redpanda/compare/703b3d42ca02517433849478dbe6b6b937b1a6b4..778fc476d86ce0e720730bca10b6fe96345ab818

dotnwat reviewed (commented):
This PR is missing a coverletter

dotnwat commented on src/v/datalake/translation/translation_probe.h:
```diff
@@ -0,0 +1,68 @@
+    // Note: Do not forget to register new causes in
+    // register_invalid_record_metric.
+    enum class invalid_record_cause {
+        /// Failed to resolve the Kafka schema for the record. This covers the
+        /// cases where magic byte is missing from the record or it references a
+        /// non-existent schema in the registry.
+        failed_kafka_schema_resolution,
+        /// Failed to translate the record data according to the schema fetches
+        /// from schema registry to an equivalent Iceberg schema.
+        failed_data_translation,
+        /// Failed to ensure the table schema matches the inferred Iceberg
+        /// schema.
```
🥳 comments

dotnwat commented on src/v/datalake/translation/translation_probe.h:
```diff
@@ -0,0 +1,68 @@
+        counter_ref(cause)++;
+    }
+
+    constexpr size_t& counter_ref(invalid_record_cause cause) {
```
i'm not sure i really get the point of constexpr here

nvartolomei replied:
https://github.com/redpanda-data/redpanda/pull/25116
